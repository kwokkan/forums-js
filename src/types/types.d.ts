
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
