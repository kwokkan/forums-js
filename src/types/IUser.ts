export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
