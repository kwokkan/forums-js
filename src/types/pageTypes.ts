import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
