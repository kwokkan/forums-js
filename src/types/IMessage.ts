import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
