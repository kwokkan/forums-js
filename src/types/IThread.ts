import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
