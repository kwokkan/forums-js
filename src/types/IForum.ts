import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
