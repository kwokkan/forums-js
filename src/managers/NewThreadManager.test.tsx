import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { act } from "react-dom/test-utils";
import { create } from "react-test-renderer";
import { IThread } from "../types/IThread";

beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
});

test("Renders without error", () => {
    jest.isolateModules(() => {
        const { NewThreadManager } = require("./NewThreadManager");

        const tree = create(
            <NewThreadManager />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});

test("onNewThread callback called", (doneCallback) => {
    jest.isolateModules(async () => {
        const mockAddThread = jest.fn((): IThread => ({
            created: 1,
            id: 20,
            messages: [],
            name: "New title"
        }));
        const mockPush = jest.fn();

        jest.doMock("../api", () => {
            return {
                __esModule: true,
                addThread: mockAddThread
            };
        });

        jest.doMock("next/router", () => {
            return {
                __esModule: true,
                useRouter: () => ({
                    push: mockPush
                })
            };
        });

        const { NewThreadManager } = await import("./NewThreadManager");

        let wrapper = mount(
            <NewThreadManager forumId={10} />
        );

        act(() => {
            wrapper.find('input[name="title"]').simulate("change", { target: { value: "New title" } });
            wrapper.find('textarea[name="message"]').simulate("change", { target: { value: "New message" } });
        });

        await act(async () => {
            wrapper.find("button").simulate("click");
        });

        setImmediate(() => {
            expect(mockAddThread).toBeCalledWith({ forumId: 10, title: "New title", message: "New message" });
            expect(mockPush).toBeCalledWith("/threads/[id]", "/threads/20");

            doneCallback();
        });
    });
});
