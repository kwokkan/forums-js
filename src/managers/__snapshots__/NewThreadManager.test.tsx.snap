// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without error 1`] = `
<div
  className="rs-panel"
>
  <div
    aria-expanded={false}
    className="rs-panel-header"
    tabIndex={-1}
  >
    <h3
      className="rs-panel-title"
    >
      New thread
    </h3>
  </div>
  <div
    className="rs-panel-body"
    role="region"
  >
    <div
      className="rs-panel-group"
    >
      <main
        className="rs-content"
      >
        <form
          className="rs-form rs-form-vertical rs-form-fluid"
          onSubmit={[Function]}
        >
          <div
            className="rs-form-group"
            role="group"
          >
            <label
              className="rs-form-control-label"
              htmlFor="title"
              id="title-control-label"
            >
              Title
            </label>
            <div
              className="rs-form-control rs-form-control-wrapper"
            >
              <input
                aria-describedby="title-help-text"
                aria-labelledby="title-control-label"
                className="rs-input"
                id="title"
                name="title"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="rs-form-group"
            role="group"
          >
            <label
              className="rs-form-control-label"
              htmlFor="message"
              id="message-control-label"
            >
              Message
            </label>
            <div
              className="mb-3 rs-form-control rs-form-control-wrapper"
            >
              <textarea
                aria-describedby="message-help-text"
                aria-labelledby="message-control-label"
                className="rs-input"
                id="message"
                name="message"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                rows={10}
                type="text"
                value=""
              />
            </div>
          </div>
        </form>
      </main>
    </div>
    <button
      className="rs-btn rs-btn-primary"
      onClick={[Function]}
      type="button"
    >
      Submit
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
        />
      </span>
    </button>
  </div>
</div>
`;
